### PR TITLE
yazi-unwrapped: auto update cargo.lock

### PR DESCRIPTION
**Pull request by XYenon**
## Description of changes

Make the update script be able to update `Cargo.lock`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ya/yazi-unwrapped/update.sh [hunk 1]
```diff
@@ -14,10 +14,10 @@ YAZI_RELEASE=$(
 
 # Get release information
 latestBuildDate=$(echo "$YAZI_RELEASE" | jq -r ".published_at")
-latestVersion=$(echo "$YAZI_RELEASE" | jq -r ".tag_name")
+tagName=$(echo "$YAZI_RELEASE" | jq -r ".tag_name")
 
 latestBuildDate="${latestBuildDate%T*}" # remove the timestamp and get the date
-latestVersion="${latestVersion:1}"      # remove first char 'v'
+latestVersion="${tagName:1}"      # remove first char 'v'
```
Comment (eljamm) [inline pkgs/by-name/ya/yazi-unwrapped/update.sh:20]: it's more readable if the comments are aligned, in my opinion

```suggestion
latestVersion="${tagName:1}"            # remove first char 'v'
```

File: pkgs/by-name/ya/yazi-unwrapped/update.sh [hunk 2]
```diff
@@ -35,7 +35,11 @@ pushd "$SCRIPT_DIR"
...
 
+# Cargo.lock
+curl -L -o 'Cargo.lock' "https://raw.githubusercontent.com/sxyazi/yazi/$tagName/Cargo.lock"
+git add 'Cargo.lock'
+
 # Hashes
-cargoHash=$(nix-prefetch "{ sha256 }: (import $NIXPKGS_DIR {}).yazi-unwrapped.cargoDeps.overrideAttrs (_: { outputHash = sha256; })")
-sed -i -E 's#\bcargoHash = ".*?"#cargoHash = "'"$cargoHash"'"#' package.nix
+# cargoHash=$(nix-prefetch "{ sha256 }: (import $NIXPKGS_DIR {}).yazi-unwrapped.cargoDeps.overrideAttrs (_: { outputHash = sha256; })")
+# sed -i -E 's#\bcargoHash = ".*?"#cargoHash = "'"$cargoHash"'"#' package.nix
```
Comment (eljamm) [inline pkgs/by-name/ya/yazi-unwrapped/update.sh:44]: Perhaps we can parse the `Cargo.lock` file and see if there are any "git+" dependencies to know whether to use `cargoLock` or `cargoHash`. What do you think?

File: pkgs/by-name/ya/yazi-unwrapped/update.sh [hunk 3]
```diff
@@ -35,7 +35,11 @@ pushd "$SCRIPT_DIR"
 # Build date
 sed -i 's#env.VERGEN_BUILD_DATE = "[^"]*"#env.VERGEN_BUILD_DATE = "'"${latestBuildDate}"'"#' package.nix
 
+# Cargo.lock
+curl -L -o 'Cargo.lock' "https://raw.githubusercontent.com/sxyazi/yazi/$tagName/Cargo.lock"
+git add 'Cargo.lock'
```
Comment (eljamm) [inline pkgs/by-name/ya/yazi-unwrapped/update.sh:40]: If the script is gonna use `git`, we must add it to the `nix-shell` packages above, right?

Comment (XYenon) [reply pkgs/by-name/ya/yazi-unwrapped/update.sh:40]: Yeah, but `maintainers/scripts/update.nix` has already used `git`, should we add `git` again?

File: pkgs/by-name/ya/yazi-unwrapped/update.sh (hunk 2 above)
Comment (XYenon) [reply pkgs/by-name/ya/yazi-unwrapped/update.sh:44]: Great idea. Find it seems easy, but switch between `cargoHash` and `cargoLock` may be difficult. I'll try to implement it.

File: pkgs/by-name/ya/yazi-unwrapped/update.sh (hunk 3 above)
Comment (eljamm) [reply pkgs/by-name/ya/yazi-unwrapped/update.sh:40]: Well, it does work without explicitly adding it, so I guess it's fine.

File: pkgs/by-name/ya/yazi-unwrapped/update.sh (hunk 2 above)
Comment (XYenon) [reply pkgs/by-name/ya/yazi-unwrapped/update.sh]: I have implemented it by ast-grep.

**Review by eljamm — APPROVED** [review]
This looks amazing. Thanks @XYenon!

I don't know if the implementation can be simplified further, but I tested this both ways and it works.

That said, you might want to add some comments to explain what each step does to make things more readable for future maintainers.

Aside from that, I think this is good to merge.

**Review by eljamm — APPROVED** [review]
LGTM, thanks!

**Comment by uncenter** [comment]
I'm not sure how I feel about this complexity being added, especially since https://github.com/notify-rs/notify is the only Git dep at the moment and will hopefully be removed (provided they make a new patch release 🙏) by the next update.

**Comment by eljamm** [comment]
That's true and I agree with you on the complexity, but I also like the idea of having a future-proof auto update script as well. It seems to be working nicely as is, so maybe we can iterate over it and improve it over time.

**Comment by XYenon** [comment]
> I'm not sure how I feel about this complexity being added, especially since [notify-rs/notify](https://github.com/notify-rs/notify) is the only Git dep at the moment and will hopefully be removed (provided they make a new patch release 🙏) by the next update.

`nix-update-script` can deal with this, actually. But not the `env.VERGEN_BUILD_DATE = "2024-04-23";`.

**Comment by eljamm** [comment]
There is a new `useFetchCargoVendor` fetcher that can handle git deps:

https://github.com/NixOS/nixpkgs/blob/45533e4b16bcdafbd28eec81cc959028e8e9982c/doc/languages-frameworks/rust.section.md?plain=1#L66-L75

Perhaps we can rewrite this PR to simply include or remove it as needed.

**Pull request closed without merging**